tests/jax/core_system_components/parameter_client_test.py

```diff
@@ -15,4 +15,410 @@
+
+    # set call counter equal to update period and verify the request was made
+    parameter_client._get_call_counter = 10
+    parameter_client.get_async()
+
+    assert parameter_client._parameters == {
+        "key_0": np.array(0, dtype=np.int32),
+        "key_1": np.array(2, dtype=np.float32),
+        "key_2": np.array(2, dtype=np.int32),
+        "key_3": np.array(4, dtype=np.int32),
+        "key_4": np.array(5, dtype=np.int32),
+        "networks-network_key_0": {"layer_0": {"weights": 1, "biases": 1}},
+        "networks-network_key_1": {"layer_0": {"weights": 2, "biases": 2}},
+        "networks-network_key_2": {"layer_0": {"weights": 3, "biases": 3}},
+        "allkey_0": np.array(0, dtype=np.int32),
+    }
+    assert parameter_client._get_call_counter == 0
+    assert parameter_client._get_future is None
+
+
+def test_set_async(parameter_client: ParameterClient) -> None:
+    """Set set async method"""
+
+    # test that set_call_counter is incremented
+    parameter_client.set_async()
+    assert parameter_client._set_call_counter == 1
+
+    # set call counter equal to update period and verify the request was made
+    parameter_client._set_call_counter = 10
+
+    # increment set parameters, call set_async and verify that
+    # parameters were set asynchronously
+    increment_set_parameters(
+        params=parameter_client._parameters, names=parameter_client._set_keys
+    )
+
+    parameter_client.set_async()
+
+    assert parameter_client._client.store._set_params == {
+        "key_0": np.array(1, dtype=np.int32),
+        "key_2": np.array(3, dtype=np.int32),
+    }
+    assert parameter_client._set_call_counter == 0
+
+    # assert that _set_future is None again after request was made
+    parameter_client.set_async()
+    assert parameter_client._set_future is None
+
+
+def test_set_and_get_async(parameter_client: ParameterClient) -> None:
+    """Set set async method"""
+
+    # test that set_and_get_call_counter is incremented
+    parameter_client.set_and_get_async()
+    assert parameter_client._set_get_call_counter == 1
+
+    # set counter equal to update period and verify the request was made
+    parameter_client._set_get_call_counter = 10
+
+    # increment set parameters
+    increment_set_parameters(
+        params=parameter_client._parameters, names=parameter_client._set_keys
+    )
+
+    parameter_client.set_and_get_async()
+
+    assert parameter_client._parameters == {
+        "key_0": np.array(1, dtype=np.int32),
+        "key_1": np.array(2, dtype=np.float32),
+        "key_2": np.array(3, dtype=np.int32),
+        "key_3": np.array(4, dtype=np.int32),
+        "key_4": np.array(5, dtype=np.int32),
+        "networks-network_key_0": {"layer_0": {"weights": 1, "biases": 1}},
+        "networks-network_key_1": {"layer_0": {"weights": 2, "biases": 2}},
+        "networks-network_key_2": {"layer_0": {"weights": 3, "biases": 3}},
+        "allkey_0": np.array(0, dtype=np.int32),
```

Comment:
You expect this 3 times maybe extract into `expected_get_params`? But I can see an argument to leave it as is too.

Reply:
I think it is a bit easier to leave it as is for now.